packages/puppeteer-core/src/bidi/Browser.ts

```
@@ -229,29 +217,33 @@ export class BidiBrowser extends Browser {
   }
 
   get connection(): BidiConnection {
-    return this.#connection;
+    // SAFETY: We only have one implementation.
+    return this.#browserCore.session.connection as BidiConnection;
   }
 
   override wsEndpoint(): string {
-    return this.#connection.url;
+    return this.connection.url;
   }
 
   override async close(): Promise<void> {
     for (const [eventName, handler] of this.#connectionEventHandlers) {
-      this.#connection.off(eventName, handler);
+      this.connection.off(eventName, handler);
     }
-    if (this.#connection.closed) {
+    if (this.connection.closed) {
       return;
     }
 
-    // `browser.close` can close connection before the response is received.
-    await this.#connection.send('browser.close', {}).catch(debugError);
-    await this.#closeCallback?.call(null);
-    this.#connection.dispose();
+    try {
+      await this.#browserCore.close();
+      await this.#closeCallback?.call(null);
+    } catch (error) {
+      // Fail silently.
```

I think we should not fail silently for closeCallback and also we should run it even if #browserCore.close() failed.  Perhaps move it to finally?

---

`closeCallback` calls `close` from looking at the location of the callback, so it's fine to fail silently here. We need to rework that logic at some point.

---

It should force close the browser process if the protocol methods failed. The callback is in ProductLauncher. Agree on refactor, not sure if this change will affect any behavior in practice, probably not.

---

The tests seem to take quite longer than usual https://github.com/puppeteer/puppeteer/actions/runs/7583926821/job/20656609557?pr=11702